tooling/metamodel-generator/src/main/java/org/hibernate/processor/annotation/AnnotationMetaEntity.java

```diff
@@ -3388,19 +3389,29 @@ private List<String> parameterTypes(ExecutableElement method) {
 	 * Workaround for a bug in Java 20/21. Should not be necessary!
 	 */
 	private String typeAsString(TypeMirror type) {
-		String result = type.toString();
-		for ( AnnotationMirror annotation : type.getAnnotationMirrors() ) {
-			final String annotationString = annotation.toString();
-			result = result
-					// if it has a space after it, we need to remove that too
-					.replace(annotationString + ' ', "")
-					// just in case it did not have a space after it
-					.replace(annotationString, "");
-		}
-		for ( AnnotationMirror annotation : type.getAnnotationMirrors() ) {
-			result = annotation.toString() + ' ' + result;
+		return typeAsString( type, true );
+	}
+
+	private String typeAsString(TypeMirror type, boolean includeAnnotations) {
+		if ( type instanceof DeclaredType dt && dt.asElement() instanceof TypeElement te ) {
+			// get the "fqcn" without any type arguments
+			String result = te.getQualifiedName().toString();
+			// add the < ? ,? ....> as necessary:
+			if ( !dt.getTypeArguments().isEmpty() ) {
+				result += dt.getTypeArguments().stream()
+						.map( arg -> typeAsString( arg, true ) )
+						.collect( Collectors.joining( ", ", "<", ">" ) );
+			}
+			if ( includeAnnotations ) {
+				for ( AnnotationMirror annotation : type.getAnnotationMirrors() ) {
+					result = annotation.toString() + ' ' + result;
+				}
+			}
+			return result;
+		}
```

Comment:
Maybe will be better to use `StringBuilder`. I guess that this will reduce amount of job for garbage collector.

```suggestion
		if ( type instanceof DeclaredType dt && dt.asElement() instanceof TypeElement te ) {
			final StringBuilder sb = new StringBuilder();
			if ( includeAnnotations ) {
				for ( AnnotationMirror annotation : type.getAnnotationMirrors() ) {
					sb.append( annotation.toString() ).append( ' ' );
				}
			}
			// get the "fqcn" without any type arguments
			sb.append( te.getQualifiedName().toString() );
			// add the < ? ,? ....> as necessary:
			if ( !dt.getTypeArguments().isEmpty() ) {
				sb.append( '<' );
				dt.getTypeArguments()
						.forEach( arg -> sb.append( typeAsString( arg, true ) ) );
				sb.append( '>' );
			}
			return sb.toString();
		}
```

Reply:
I agree.